lineage/LineageTree.py

```diff
@@ -44,234 +27,137 @@ def __init__(self, pi, T, E, desired_num_cells, desired_experiment_time, prune_c
+                break
 
-        indices_of_cells_in_state {list}: Holding the indexes of the cells being in the given state.
+    def output_assign_obs(self, state):
         """
-        cells_in_state = []  # a list holding cells in the same state
-        list_of_tuples_of_obs = []
-        indices_of_cells_in_state = []
-        for cell in self.pruned_lin_list:
-            if cell.state == state:  # if the cell is in the given state...
-                cells_in_state.append(cell)  # append them to a list
-                list_of_tuples_of_obs.append(cell.obs)
-                indices_of_cells_in_state.append(
-                    self.pruned_lin_list.index(cell))
-        # gets the number of cells in the list
-        num_cells_in_state = len(cells_in_state)
-
-        return num_cells_in_state, cells_in_state, list_of_tuples_of_obs, indices_of_cells_in_state
-
-    def _full_assign_obs(self, state):
-        """ Observation assignment give a state.
+        Observation assignment give a state.
         Given the lineageTree object and the intended state, this function assigns the corresponding observations
         comming from specific distributions for that state.
 
         Args:
         -----
         state {Int}: The number assigned to a state.
 
-        Returns the same outputs as the `_get_pruned_state_count()`.
         """
-        num_cells_in_state, cells_in_state, indices_of_cells_in_state = self._get_full_state_count(
-            state)
-        list_of_tuples_of_obs = self.E[state].rvs(size=num_cells_in_state)
-
-        assert len(cells_in_state) == len(
-            list_of_tuples_of_obs) == num_cells_in_state
-
+        cells_in_state = [cell for cell in self.full_lineage if cell.state == state]
+        list_of_tuples_of_obs = self.E[state].rvs(size=len(cells_in_state))
+        assert len(cells_in_state) == len(list_of_tuples_of_obs)
         for i, cell in enumerate(cells_in_state):
             cell.obs = list_of_tuples_of_obs[i]
 
-        return num_cells_in_state, cells_in_state, list_of_tuples_of_obs, indices_of_cells_in_state
-
-    def _get_parents_for_level(self, level):
-        """ get the parents's index of a generation in the population list.
+    def censor_lineage(self):
+        """
+        This function removes those cells that are intended to be remove
+        from the output binary tree based on emissions.
+        It takes in LineageTree object, walks through all the cells in the output binary tree,
+        applies the pruning to each cell that is supposed to be removed,
+        and returns the censord list of cells.
+        """
+        self.output_lineage = []
+        for idx, cell in enumerate(self.full_lineage):
+            if self.censor_condition == 0:
+                self.output_lineage = self.full_lineage
+                break
+            elif self.censor_condition == 1:
+                if fate_censor_rule(cell):
```

Comment:
Rather than find and copy the subtrees over and over, why not have a method `censor_cell_and_daughters()`, then have it be a recursive function that walks down the tree until a cell is a leaf?

Reply:
Yea that would avoid the huge number of for loops, you're right

Reply:
We most likely have the recursions written as functions already; I'll try to see if I can implement this using those. I can also do this on a separate branch, or on this one if you'd like.

Reply:
We can merge then make this change.